File: app/src/main/java/com/example/myapplication/ui/screens/WatchlistScreen.kt
package com.example.myapplication.ui.screens

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.KeyboardArrowDown
import androidx.compose.material.icons.rounded.KeyboardArrowUp
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.example.myapplication.data.model.Media
import com.example.myapplication.data.model.WatchCategory
import com.example.myapplication.data.model.WatchStatus
import com.example.myapplication.data.model.WatchlistItem
import com.example.myapplication.data.model.statuses
import java.util.concurrent.TimeUnit

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun WatchlistScreen(
    viewModel: WatchlistViewModel,
    category: WatchCategory,
    onMediaClick: (Media) -> Unit
) {
    val watchlist by viewModel.watchlist(category).collectAsState()
    val availableStatuses = category.statuses()
    val expandedStates = remember(category) {
        mutableStateMapOf<WatchStatus, Boolean>().apply {
            availableStatuses.forEach { put(it, true) }
        }
    }
    val staleSectionLabel = "Pas regardé depuis un moment"
    val staleItems = remember(watchlist, category) {
        if (category == WatchCategory.FILMS) {
            emptyList()
        } else {
            watchlist.filter { it.isStale() }
        }
    }
    val regularItems = remember(watchlist, staleItems) {
        watchlist.filterNot { staleItems.contains(it) }
    }
    val staleExpanded = remember(category) { mutableStateOf(true) }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text(category.label, fontWeight = FontWeight.Bold) },
            )
        }
    ) { padding ->
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding),
            contentPadding = PaddingValues(bottom = 16.dp)
        ) {
            if (category != WatchCategory.FILMS) {
                item {
                    Column(modifier = Modifier.fillMaxWidth()) {
                        SectionHeader(
                            title = staleSectionLabel,
                            expanded = staleExpanded.value,
                            onToggle = { staleExpanded.value = !staleExpanded.value }
                        )
                        if (staleExpanded.value) {
                            if (staleItems.isEmpty()) {
                                Box(
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .padding(horizontal = 16.dp, vertical = 12.dp),
                                    contentAlignment = Alignment.Center
                                ) {
                                    Text("Aucun titre dans cette sous-liste", style = MaterialTheme.typography.bodyMedium)
                                }
                            } else {
                                SectionGrid(
                                    items = staleItems,
                                    onMediaClick = onMediaClick
                                )
                            }
                        }
                    }
                }
            }

            items(availableStatuses.size) { index ->
                val status = availableStatuses[index]
                val sectionItems = regularItems.filter { it.status == status }
                val expanded = expandedStates[status] ?: true

                Column(modifier = Modifier.fillMaxWidth()) {
                    SectionHeader(
                        title = status.label,
                        expanded = expanded,
                        onToggle = { expandedStates[status] = !expanded }
                    )
                    if (expanded) {
                        if (sectionItems.isEmpty()) {
                            Box(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(horizontal = 16.dp, vertical = 12.dp),
                                contentAlignment = Alignment.Center
                            ) {
                                Text("Aucun titre dans cette sous-liste", style = MaterialTheme.typography.bodyMedium)
                            }
                        } else {
                            SectionGrid(
                                items = sectionItems,
                                onMediaClick = onMediaClick
                            )
                        }
                    }
                }
            }
        }
    }
}

private fun WatchlistItem.isStale(): Boolean {
    val watchedAt = lastWatchedAt ?: return false
    if (watchedEpisodes <= 0) return false
    val thirtyDaysMs = TimeUnit.DAYS.toMillis(30)
    return System.currentTimeMillis() - watchedAt > thirtyDaysMs
}

@Composable
private fun SectionHeader(
    title: String,
    expanded: Boolean,
    onToggle: () -> Unit
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onToggle)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(text = title, style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.Bold)
        Spacer(Modifier.weight(1f))
        Icon(
            imageVector = if (expanded) Icons.Rounded.KeyboardArrowUp else Icons.Rounded.KeyboardArrowDown,
            contentDescription = if (expanded) "Réduire" else "Développer"
        )
    }
}

@Composable
private fun SectionGrid(
    items: List<WatchlistItem>,
    onMediaClick: (Media) -> Unit
) {
    Column(modifier = Modifier.padding(horizontal = 8.dp)) {
        items.chunked(3).forEach { rowItems ->
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 8.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                rowItems.forEach { item ->
                    WatchlistCard(
                        item = item,
                        onMediaClick = onMediaClick
                    )
                }
                if (rowItems.size < 3) {
                    repeat(3 - rowItems.size) {
                        Spacer(modifier = Modifier.weight(1f))
                    }
                }
            }
        }
    }
}

@Composable
private fun RowScope.WatchlistCard(
    item: WatchlistItem,
    onMediaClick: (Media) -> Unit
) {
    val media = item.media
    val progressState = item.progressState()
    MediaCard(
        media = media,
        onClick = { onMediaClick(media) },
        modifier = Modifier.weight(1f),
        showTitle = false,
        bottomContent = {
            if (progressState != null) {
                LinearProgressIndicator(
                    progress = { progressState.progress },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(6.dp)
                        .align(Alignment.BottomCenter),
                    color = progressState.color,
                    trackColor = Color.Black.copy(alpha = 0.35f)
                )
            }
        }
    )
}

private data class ProgressState(val progress: Float, val color: Color)

private fun WatchlistItem.progressState(): ProgressState? {
    if (status == WatchStatus.NOT_STARTED || status == WatchStatus.NOT_WATCHED) {
        return null
    }

    return when (status) {
        WatchStatus.IN_PROGRESS -> {
            val progress = if (totalEpisodes > 0) {
                watchedEpisodes.coerceAtMost(totalEpisodes).toFloat() / totalEpisodes.toFloat()
            } else {
                0f
            }
            ProgressState(progress.coerceIn(0f, 1f), Color(0xFFFFC107))
        }
        WatchStatus.UP_TO_DATE -> ProgressState(1f, Color(0xFF4CAF50))
        WatchStatus.COMPLETED -> ProgressState(1f, Color(0xFF9C27B0))
        WatchStatus.WATCHED -> ProgressState(1f, Color(0xFF9C27B0))
        else -> null
    }
}

File: flutter_app/lib/src/presentation/screens/watchlist_screen.dart
import 'package:flutter/material.dart';

import '../../data/models/media_models.dart';
import '../../data/repositories/media_repository.dart';
import '../theme/slide_up_route.dart';
import '../widgets/media_card.dart';
import 'details_screen.dart';

class WatchlistScreen extends StatefulWidget {
  const WatchlistScreen({
    super.key,
    required this.repository,
    required this.category,
  });

  final MediaRepository repository;
  final WatchCategory category;

  @override
  State<WatchlistScreen> createState() => _WatchlistScreenState();
}

class _WatchlistScreenState extends State<WatchlistScreen> {
  late Future<_WatchlistSections> _future;
  final Map<WatchStatus, bool> _expanded = {};
  bool _staleExpanded = true;

  @override
  void initState() {
    super.initState();
    _reload();
    widget.repository.watchlistVersion.addListener(_onWatchlistChanged);
    for (final s in WatchStatus.values) {
      _expanded[s] = true;
    }
  }

  @override
  void dispose() {
    widget.repository.watchlistVersion.removeListener(_onWatchlistChanged);
    super.dispose();
  }

  void _onWatchlistChanged() {
    if (!mounted) return;
    _reload();
  }

  void _reload() {
    setState(() {
      _future = _loadSections();
    });
  }

  Future<_WatchlistSections> _loadSections() async {
    final items = await widget.repository.getWatchlist(widget.category);
    return _WatchlistSections.fromItems(widget.category, items);
  }

  @override
  Widget build(BuildContext context) {
    final statuses = _statusesForCategory(widget.category);
    return Scaffold(
      appBar: AppBar(
        title: Text(
          widget.category.label,
          style: const TextStyle(fontWeight: FontWeight.bold),
        ),
      ),
      body: FutureBuilder<_WatchlistSections>(
        future: _future,
        builder: (context, snapshot) {
          if (snapshot.connectionState != ConnectionState.done) {
            return const Center(child: CircularProgressIndicator());
          }
          final sections = snapshot.data ?? _WatchlistSections.empty();

          return RefreshIndicator(
            onRefresh: () async => _reload(),
            child: ListView(
              padding: const EdgeInsets.only(bottom: 16),
              children: [
                if (widget.category != WatchCategory.films) ...[
                  _SectionHeader(
                    title: 'Pas regardé depuis un moment',
                    expanded: _staleExpanded,
                    onToggle: () =>
                        setState(() => _staleExpanded = !_staleExpanded),
                  ),
                  if (_staleExpanded)
                    sections.staleItems.isEmpty
                        ? _emptySection()
                        : _SectionGrid(
                            items: sections.staleItems,
                            onTap: _openDetails,
                          ),
                ],
                for (final status in statuses) ...[
                  _SectionHeader(
                    title: status.label,
                    expanded: _expanded[status] ?? true,
                    onToggle: () => setState(
                      () => _expanded[status] = !(_expanded[status] ?? true),
                    ),
                  ),
                  if (_expanded[status] ?? true)
                    ((sections.byStatus[status] ?? const <WatchlistItem>[])
                            .isEmpty)
                        ? _emptySection()
                        : _SectionGrid(
                            items: sections.byStatus[status]!,
                            onTap: _openDetails,
                          ),
                ],
              ],
            ),
          );
        },
      ),
    );
  }

  Widget _emptySection() {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
      child: Center(
        child: Text(
          'Aucun titre dans cette sous-liste',
          style: Theme.of(context).textTheme.bodyMedium,
        ),
      ),
    );
  }

  Future<void> _openDetails(WatchlistItem item) async {
    await Navigator.of(context).push(
      SlideUpRoute<void>(
        builder: (_) =>
            DetailsScreen(repository: widget.repository, media: item.media),
      ),
    );
    _reload();
  }

  List<WatchStatus> _statusesForCategory(WatchCategory category) {
    return switch (category) {
      WatchCategory.films => const [
        WatchStatus.notWatched,
        WatchStatus.watched,
      ],
      _ => const [
        WatchStatus.notStarted,
        WatchStatus.inProgress,
        WatchStatus.upToDate,
        WatchStatus.completed,
      ],
    };
  }
}

class _WatchlistSections {
  const _WatchlistSections({required this.staleItems, required this.byStatus});

  final List<WatchlistItem> staleItems;
  final Map<WatchStatus, List<WatchlistItem>> byStatus;

  factory _WatchlistSections.empty() =>
      const _WatchlistSections(staleItems: [], byStatus: {});

  factory _WatchlistSections.fromItems(
    WatchCategory category,
    List<WatchlistItem> items,
  ) {
    final staleItems = <WatchlistItem>[];
    final byStatus = <WatchStatus, List<WatchlistItem>>{};
    final staleCutoffMillis = DateTime.now()
        .subtract(const Duration(days: 30))
        .millisecondsSinceEpoch;

    for (final item in items) {
      final watchedAt = item.lastWatchedAt;
      final isStale =
          category != WatchCategory.films &&
          watchedAt != null &&
          item.watchedEpisodes > 0 &&
          watchedAt < staleCutoffMillis;
      if (isStale) {
        staleItems.add(item);
        continue;
      }
      (byStatus[item.status] ??= <WatchlistItem>[]).add(item);
    }

    return _WatchlistSections(staleItems: staleItems, byStatus: byStatus);
  }
}

class _SectionHeader extends StatelessWidget {
  const _SectionHeader({
    required this.title,
    required this.expanded,
    required this.onToggle,
  });

  final String title;
  final bool expanded;
  final VoidCallback onToggle;

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onToggle,
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
        child: Row(
          children: [
            Text(
              title,
              style: Theme.of(context).textTheme.titleMedium
                  ?.copyWith(fontWeight: FontWeight.bold),
            ),
            const Spacer(),
            Icon(
              expanded
                  ? Icons.keyboard_arrow_up_rounded
                  : Icons.keyboard_arrow_down_rounded,
            ),
          ],
        ),
      ),
    );
  }
}

class _SectionGrid extends StatelessWidget {
  const _SectionGrid({required this.items, required this.onTap});

  final List<WatchlistItem> items;
  final Future<void> Function(WatchlistItem) onTap;

  @override
  Widget build(BuildContext context) {
    final rows = <List<WatchlistItem>>[];
    for (var i = 0; i < items.length; i += 3) {
      rows.add(items.sublist(i, i + 3 > items.length ? items.length : i + 3));
    }
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 8),
      child: Column(
        children: rows.map((row) {
          return Padding(
            padding: const EdgeInsets.only(bottom: 8),
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                ...row.map(
                  (item) => Expanded(
                    child: Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 4),
                      child: _WatchlistCard(item: item, onTap: onTap),
                    ),
                  ),
                ),
                if (row.length < 3)
                  ...List.generate(
                    3 - row.length,
                    (_) => const Expanded(child: SizedBox()),
                  ),
              ],
            ),
          );
        }).toList(),
      ),
    );
  }
}

class _WatchlistCard extends StatelessWidget {
  const _WatchlistCard({required this.item, required this.onTap});

  final WatchlistItem item;
  final Future<void> Function(WatchlistItem) onTap;

  Color? _progressColor() {
    return switch (item.status) {
      WatchStatus.inProgress => const Color(0xFFFFC107),
      WatchStatus.upToDate => const Color(0xFF4CAF50),
      WatchStatus.completed => const Color(0xFF9C27B0),
      WatchStatus.watched => const Color(0xFF9C27B0),
      _ => null,
    };
  }

  double _progress() {
    if (item.status == WatchStatus.inProgress) {
      if (item.totalEpisodes <= 0) return 0;
      return (item.watchedEpisodes / item.totalEpisodes).clamp(0.0, 1.0);
    }
    return 1.0;
  }

  @override
  Widget build(BuildContext context) {
    final progressColor = _progressColor();
    final showProgress =
        item.media.mediaType != MediaType.movie &&
        item.status != WatchStatus.notStarted &&
        item.status != WatchStatus.notWatched &&
        progressColor != null;

    return MediaCard(
      posterPath: item.media.posterPath,
      title: item.media.title,
      onTap: () => onTap(item),
      bottomContent: showProgress
          ? Positioned(
              bottom: 0,
              left: 0,
              right: 0,
              child: LinearProgressIndicator(
                value: _progress(),
                minHeight: 6,
                color: progressColor,
                backgroundColor: Colors.black.withValues(alpha: 0.35),
              ),
            )
          : null,
    );
  }
}
